File: iot_monitor/sensor_types.py
"""
sensor_types.py — Los 5 tipos de sensores IoT simulados.

Cada clase hereda de SensorClient e implementa:
  - sensor_type : tipo según protocolo SIMP
  - unit        : unidad de medición
  - measure()   : generación realista de datos simulados

Sensores implementados:
  1. TemperatureSensor   — TEMPERATURE  (Celsius)
  2. HumiditySensor      — HUMIDITY     (Percent)
  3. PressureSensor      — PRESSURE     (hPa)
  4. VibrationSensor     — VIBRATION    (g)
  5. EnergySensor        — ENERGY       (kWh)
"""

import random
import math
import time
from sensor_client import SensorClient


# ─────────────────────────────────────────────────────────────
# 1. Sensor de Temperatura
# ─────────────────────────────────────────────────────────────
class TemperatureSensor(SensorClient):
    """
    Simula un sensor de temperatura industrial.
    Rango normal: 15°C – 40°C
    Umbral de alerta: < -10°C  o  > 90°C

    Modelo: oscilación sinusoidal diaria + ruido gaussiano.
    Ocasionalmente genera picos para probar las alertas.
    """

    def __init__(self, sensor_id: str, token: str, **kwargs):
        super().__init__(sensor_id, token, **kwargs)
        self._base_temp   = random.uniform(20.0, 30.0)
        self._spike_prob  = 0.03   # 3% de probabilidad de pico
        self._start_time  = time.time()

    @property
    def sensor_type(self) -> str:
        return "TEMPERATURE"

    @property
    def unit(self) -> str:
        return "Celsius"

    def measure(self) -> float:
        # Oscilación sinusoidal: ciclo de 24 h simulado en 2 min
        elapsed  = time.time() - self._start_time
        period   = 120.0                        # 2 minutos = "1 día simulado"
        osc      = 8.0 * math.sin(2 * math.pi * elapsed / period)
        noise    = random.gauss(0, 0.5)
        value    = self._base_temp + osc + noise

        # Spike ocasional (para probar sistema de alertas)
        if random.random() < self._spike_prob:
            value += random.uniform(55.0, 70.0)
            self.log.warning(f"Pico de temperatura simulado: {value:.2f}°C")

        return max(-20.0, min(120.0, value))   # clamping físico


# ─────────────────────────────────────────────────────────────
# 2. Sensor de Humedad
# ─────────────────────────────────────────────────────────────
class HumiditySensor(SensorClient):
    """
    Simula un sensor de humedad relativa.
    Rango normal: 30% – 70%
    Umbral de alerta: > 95%

    Modelo: tendencia lenta + ruido + eventos de lluvia simulados.
    """

    def __init__(self, sensor_id: str, token: str, **kwargs):
        super().__init__(sensor_id, token, **kwargs)
        self._base_humidity = random.uniform(40.0, 60.0)
        self._trend         = 0.0
        self._rain_event    = False
        self._rain_duration = 0

    @property
    def sensor_type(self) -> str:
        return "HUMIDITY"

    @property
    def unit(self) -> str:
        return "Percent"

    def measure(self) -> float:
        # Evento de lluvia: aumenta humedad bruscamente
        if not self._rain_event and random.random() < 0.02:
            self._rain_event    = True
            self._rain_duration = random.randint(3, 8)
            self.log.info("Evento de lluvia simulado iniciado")

        if self._rain_event:
            self._trend = random.uniform(2.0, 5.0)
            self._rain_duration -= 1
            if self._rain_duration <= 0:
                self._rain_event = False
                self._trend      = -random.uniform(0.5, 1.5)  # secado gradual
        else:
            # Leve deriva aleatoria
            self._trend += random.gauss(0, 0.1)
            self._trend  = max(-0.5, min(0.5, self._trend))

        noise = random.gauss(0, 0.3)
        self._base_humidity = max(10.0, min(98.0,
                                    self._base_humidity + self._trend + noise))
        return self._base_humidity


# ─────────────────────────────────────────────────────────────
# 3. Sensor de Presión Atmosférica
# ─────────────────────────────────────────────────────────────
class PressureSensor(SensorClient):
    """
    Simula un sensor de presión atmosférica.
    Rango normal: 980 hPa – 1030 hPa
    Umbral de alerta: < 800 hPa  o  > 1100 hPa

    Modelo: random walk suave (presión cambia lentamente).
    """

    def __init__(self, sensor_id: str, token: str, **kwargs):
        super().__init__(sensor_id, token, **kwargs)
        self._pressure = random.uniform(1000.0, 1020.0)

    @property
    def sensor_type(self) -> str:
        return "PRESSURE"

    @property
    def unit(self) -> str:
        return "hPa"

    def measure(self) -> float:
        # Random walk con tendencia de retorno al valor base
        mean_reversion = (1013.25 - self._pressure) * 0.01
        delta = random.gauss(mean_reversion, 0.3)
        self._pressure = max(750.0, min(1100.0, self._pressure + delta))
        return self._pressure


# ─────────────────────────────────────────────────────────────
# 4. Sensor de Vibración Mecánica
# ─────────────────────────────────────────────────────────────
class VibrationSensor(SensorClient):
    """
    Simula un sensor de vibración en maquinaria industrial.
    Rango normal: 0 g – 0.5 g
    Umbral de alerta: > 2.0 g (fallo mecánico inminente)

    Modelo: nivel base bajo + picos de vibración por fallas simuladas.
    """

    def __init__(self, sensor_id: str, token: str, **kwargs):
        super().__init__(sensor_id, token, **kwargs)
        self._base_vibration = random.uniform(0.01, 0.1)
        self._failure_prob   = 0.04   # 4% de probabilidad de fallo

    @property
    def sensor_type(self) -> str:
        return "VIBRATION"

    @property
    def unit(self) -> str:
        return "g"

    def measure(self) -> float:
        # Vibración base (ruido de máquina normal)
        noise = abs(random.gauss(0, 0.02))
        value = self._base_vibration + noise

        # Fallo mecánico simulado
        if random.random() < self._failure_prob:
            spike = random.uniform(1.5, 4.0)
            value += spike
            self.log.warning(f"Fallo mecánico simulado: {value:.3f}g")

        return max(0.0, value)


# ─────────────────────────────────────────────────────────────
# 5. Sensor de Consumo Energético
# ─────────────────────────────────────────────────────────────
class EnergySensor(SensorClient):
    """
    Simula un sensor de consumo energético (medidor inteligente).
    Rango normal: 0.5 kWh – 15 kWh
    Umbral de alerta: > 50 kWh (consumo anormal)

    Modelo: patrón de consumo diario realista (picos mañana/tarde).
    """

    def __init__(self, sensor_id: str, token: str, **kwargs):
        super().__init__(sensor_id, token, **kwargs)
        self._base_consumption = random.uniform(2.0, 8.0)
        self._start_time       = time.time()
        self._overload_prob    = 0.02

    @property
    def sensor_type(self) -> str:
        return "ENERGY"

    @property
    def unit(self) -> str:
        return "kWh"

    def measure(self) -> float:
        # Patrón de consumo simulado: dos picos en el "día"
        elapsed = time.time() - self._start_time
        period  = 120.0   # 2 min = "1 día simulado"
        t       = (elapsed % period) / period   # 0.0 a 1.0

        # Pico matutino (~25%) y vespertino (~75%)
        morning   = math.exp(-((t - 0.25) ** 2) / 0.005)
        afternoon = math.exp(-((t - 0.75) ** 2) / 0.005)
        pattern   = self._base_consumption * (1 + 1.5 * morning + 2.0 * afternoon)

        noise = random.gauss(0, 0.2)
        value = max(0.1, pattern + noise)

        # Sobrecarga simulada
        if random.random() < self._overload_prob:
            value += random.uniform(40.0, 55.0)
            self.log.warning(f"Sobrecarga energética simulada: {value:.2f} kWh")

        return value

File: iot_monitor/SensorTypes.java
package iot;

import java.util.Random;
import java.util.logging.Logger;

/**
 * SensorTypes.java — Implementaciones de los 5 tipos de sensores IoT.
 *
 * Cada clase interna estática hereda de SensorBase e implementa
 * getSensorType(), getUnit() y measure() con datos simulados realistas.
 *
 * Sensores implementados:
 *   1. TemperatureSensor  — TEMPERATURE (Celsius)
 *   2. HumiditySensor     — HUMIDITY    (Percent)
 *   3. PressureSensor     — PRESSURE    (hPa)
 *   4. VibrationSensor    — VIBRATION   (g)
 *   5. EnergySensor       — ENERGY      (kWh)
 */
public class SensorTypes {

    private static final Random RNG = new Random();

    // ─────────────────────────────────────────────────────────
    // 1. Sensor de Temperatura
    // ─────────────────────────────────────────────────────────
    public static class TemperatureSensor extends SensorBase {

        private final double baseTemp;
        private final long   startMs;
        private static final double SPIKE_PROB = 0.03;

        public TemperatureSensor(String id, String token,
                                  String host, int port) {
            super(id, token, host, port, 5.0);
            this.baseTemp = 20.0 + RNG.nextDouble() * 10.0;
            this.startMs  = System.currentTimeMillis();
        }

        @Override public String getSensorType() { return SimpProtocol.SENSOR_TEMP; }
        @Override public String getUnit()        { return "Celsius"; }

        @Override
        public double measure() {
            // Oscilación sinusoidal: ciclo de 2 min = "1 día simulado"
            double elapsed = (System.currentTimeMillis() - startMs) / 1000.0;
            double osc     = 8.0 * Math.sin(2 * Math.PI * elapsed / 120.0);
            double noise   = RNG.nextGaussian() * 0.5;
            double value   = baseTemp + osc + noise;

            // Pico ocasional para probar alertas
            if (RNG.nextDouble() < SPIKE_PROB) {
                double spike = 55.0 + RNG.nextDouble() * 15.0;
                value += spike;
                log.warning(String.format("Pico de temperatura: %.2f°C", value));
            }
            return Math.max(-20.0, Math.min(120.0, value));
        }
    }

    // ─────────────────────────────────────────────────────────
    // 2. Sensor de Humedad
    // ─────────────────────────────────────────────────────────
    public static class HumiditySensor extends SensorBase {

        private double humidity;
        private double trend     = 0.0;
        private int    rainLeft  = 0;

        public HumiditySensor(String id, String token,
                               String host, int port) {
            super(id, token, host, port, 7.0);
            this.humidity = 40.0 + RNG.nextDouble() * 20.0;
        }

        @Override public String getSensorType() { return SimpProtocol.SENSOR_HUMIDITY; }
        @Override public String getUnit()        { return "Percent"; }

        @Override
        public double measure() {
            if (rainLeft == 0 && RNG.nextDouble() < 0.02) {
                rainLeft = 3 + RNG.nextInt(6);
                log.info("Evento de lluvia simulado iniciado");
            }

            if (rainLeft > 0) {
                trend = 2.0 + RNG.nextDouble() * 3.0;
                rainLeft--;
                if (rainLeft == 0) trend = -(0.5 + RNG.nextDouble());
            } else {
                trend += RNG.nextGaussian() * 0.1;
                trend  = Math.max(-0.5, Math.min(0.5, trend));
            }

            humidity = Math.max(10.0, Math.min(98.0,
                       humidity + trend + RNG.nextGaussian() * 0.3));
            return humidity;
        }
    }

    // ─────────────────────────────────────────────────────────
    // 3. Sensor de Presión Atmosférica
    // ─────────────────────────────────────────────────────────
    public static class PressureSensor extends SensorBase {

        private double pressure;
        private static final double MEAN = 1013.25;

        public PressureSensor(String id, String token,
                               String host, int port) {
            super(id, token, host, port, 10.0);
            this.pressure = 1000.0 + RNG.nextDouble() * 20.0;
        }

        @Override public String getSensorType() { return SimpProtocol.SENSOR_PRESSURE; }
        @Override public String getUnit()        { return "hPa"; }

        @Override
        public double measure() {
            // Random walk con reversión a la media
            double reversion = (MEAN - pressure) * 0.01;
            double delta     = reversion + RNG.nextGaussian() * 0.3;
            pressure = Math.max(750.0, Math.min(1100.0, pressure + delta));
            return pressure;
        }
    }

    // ─────────────────────────────────────────────────────────
    // 4. Sensor de Vibración
    // ─────────────────────────────────────────────────────────
    public static class VibrationSensor extends SensorBase {

        private final double baseVibration;
        private static final double FAILURE_PROB = 0.04;

        public VibrationSensor(String id, String token,
                                String host, int port) {
            super(id, token, host, port, 3.0);
            this.baseVibration = 0.01 + RNG.nextDouble() * 0.09;
        }

        @Override public String getSensorType() { return SimpProtocol.SENSOR_VIBRATION; }
        @Override public String getUnit()        { return "g"; }

        @Override
        public double measure() {
            double noise = Math.abs(RNG.nextGaussian() * 0.02);
            double value = baseVibration + noise;

            if (RNG.nextDouble() < FAILURE_PROB) {
                double spike = 1.5 + RNG.nextDouble() * 2.5;
                value += spike;
                log.warning(String.format("Fallo mecánico simulado: %.3fg", value));
            }
            return Math.max(0.0, value);
        }
    }

    // ─────────────────────────────────────────────────────────
    // 5. Sensor de Consumo Energético
    // ─────────────────────────────────────────────────────────
    public static class EnergySensor extends SensorBase {

        private final double baseConsumption;
        private final long   startMs;
        private static final double OVERLOAD_PROB = 0.02;

        public EnergySensor(String id, String token,
                             String host, int port) {
            super(id, token, host, port, 8.0);
            this.baseConsumption = 2.0 + RNG.nextDouble() * 6.0;
            this.startMs         = System.currentTimeMillis();
        }

        @Override public String getSensorType() { return SimpProtocol.SENSOR_ENERGY; }
        @Override public String getUnit()        { return "kWh"; }

        @Override
        public double measure() {
            double elapsed  = (System.currentTimeMillis() - startMs) / 1000.0;
            double t        = (elapsed % 120.0) / 120.0;

            // Picos matutino y vespertino
            double morning   = Math.exp(-Math.pow(t - 0.25, 2) / 0.005);
            double afternoon = Math.exp(-Math.pow(t - 0.75, 2) / 0.005);
            double pattern   = baseConsumption * (1 + 1.5 * morning + 2.0 * afternoon);
            double value     = Math.max(0.1, pattern + RNG.nextGaussian() * 0.2);

            if (RNG.nextDouble() < OVERLOAD_PROB) {
                value += 40.0 + RNG.nextDouble() * 15.0;
                log.warning(String.format("Sobrecarga energética: %.2f kWh", value));
            }
            return value;
        }
    }

    private SensorTypes() {}   // utilidades estáticas — no instanciar
}
